src/AzerothWarsCSharp.Source/Researches/Ironforge/TitanforgeArtifact.cs

```diff
@@ -1,31 +1,33 @@
-// using AzerothWarsCSharp.MacroTools.FactionSystem;
-//
-// using static War3Api.Common;  using static AzerothWarsCSharp.MacroTools.GeneralHelpers;
-// {
-//   public static class TitanForgeArtifact
-//   {
-//     private static readonly int RESEARCH_ID = FourCC("R08K");
-//
-//     private static void Research()
-//     {
-//       var heldItem = UnitItemInSlot(GetTriggerUnit(), 0);
-//       Artifact heldArtifact = Artifact.GetFromTypeId(GetItemTypeId(heldItem));
-//       if (heldItem != null && heldArtifact != null && heldArtifact.Titanforged == false)
-//       {
-//         heldArtifact.Titanforge();
-//       }
-//       else
-//       {
-//         GetTriggerPlayer().AdjustPlayerState(PLAYER_STATE_RESOURCE_GOLD, 1000);
-//         GetTriggerPlayer().AdjustPlayerState(PLAYER_STATE_RESOURCE_LUMBER, 750);
-//       }
-//
-//       SetPlayerTechResearched(GetTriggerPlayer(), RESEARCH_ID, 1);
-//     }
-//
-//     public static void Setup()
-//     {
-//       RegisterResearchFinishedAction(RESEARCH_ID, Research);
-//     }
-//   }
-// }
\ No newline at end of file
+ using AzerothWarsCSharp.MacroTools.FactionSystem;
+ using static War3Api.Common;  
+ using WCSharp.Events;
+ using AzerothWarsCSharp.MacroTools.ArtifactSystem;
+ namespace AzerothWarsCSharp.Source.Researches.Ironforge
+ {
+   public static class TitanForgeArtifact
+   {
+     private static readonly int RESEARCH_ID = Constants.UPGRADE_R08K_TITANFORGE_ARTIFACT_IRONFORGE;
+
+     private static void Research()
+     {
+       var heldItem = UnitItemInSlot(GetTriggerUnit(), 0);
+       Artifact heldArtifact = ArtifactManager.GetFromTypeId(GetItemTypeId(heldItem));
```

Comment:
You should check if `heldItem` is not null before trying to set `heldArtifact`, since if `heldItem` is null then it's definitely not an Artifact.
Negligible change, feel free to skip if you're not interested.